※ PolynomialFeaturesを使用しても、必ず性能が向上するわけではない。不要な変数を増やしすぎると、<strong>過学習（Overfitting）</strong>が発生する可能性がある。

<br>

## # 線形回帰モデルの学習・評価

<br>

#### ① ライブラリの読み込み

In [2]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

<br>

#### ② サンプルデータの作成

勉強時間と点数のデータを作成する。

In [3]:
df = pd.DataFrame({
    '勉強時間': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    '点数':    [45, 48, 53, 61, 72, 78, 83, 87, 88, 89]
})

display(df)

,勉強時間,点数
0,1,45
1,2,48
2,3,53
3,4,61
4,5,72
5,6,78
6,7,83
7,8,87
8,9,88
9,10,89


<br>

#### ③ 説明変数（X）・目的変数（y）の作成

In [4]:
X = df[['勉強時間']]
y = df['点数']

<br>

#### ④ 学習データ・テストデータの分割

データを学習用80%・テスト用20%に分割する。

In [5]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

<br>

#### ⑤ 一般線形回帰モデルの学習・予測

In [6]:
lr = LinearRegression()

lr.fit(X_train, y_train)

y_pred = lr.predict(X_test)

<br>

#### ⑥ モデルの評価

R²・MAE・MSE・RMSEを用いてモデル性能を確認する。

In [7]:
print("===== 一般線形回帰 =====")
print("R²  :", r2_score(y_test, y_pred))
print("MAE :", mean_absolute_error(y_test, y_pred))
print("MSE :", mean_squared_error(y_test, y_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))

===== 一般線形回帰 =====
R²  : 0.97609375
MAE : 3.0
MSE : 9.5625
RMSE: 3.092329219213245


<br>

#### ⑦ PolynomialFeaturesによる派生変数の生成

PolynomialFeaturesを用いて、**勉強時間から「勉強時間」と「勉強時間²」の2つの特徴量を生成**する。

In [8]:
poly = PolynomialFeatures(
    degree=2,
    include_bias=False
)

X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

print("===== 派生変数の生成 =====")
print(poly.get_feature_names_out(['勉強時間']))

===== 派生変数の生成 =====
['勉強時間' '勉強時間^2']


<br>

#### ⑧ 派生変数を用いた線形回帰モデルの学習・予測

**勉強時間・勉強時間²**を用いて、線形回帰モデルを学習し予測する。

In [9]:
lr_poly = LinearRegression()

lr_poly.fit(X_train_poly, y_train)

y_pred_poly = lr_poly.predict(X_test_poly)

<br>

#### ⑨ 派生変数追加後のモデル評価

R²・MAE・MSE・RMSEを用いて、PolynomialFeatures適用後のモデル性能を確認する。

In [10]:
print("===== PolynomialFeatures 線形回帰 =====")
print("R²  :", r2_score(y_test, y_pred_poly))
print("MAE :", mean_absolute_error(y_test, y_pred_poly))
print("MSE :", mean_squared_error(y_test, y_pred_poly))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_poly)))

===== PolynomialFeatures 線形回帰 =====
R²  : 0.9945061391427036
MAE : 1.2786885245901622
MSE : 2.197544342918561
RMSE: 1.4824116644571308


<br>

#### ⑩ PolynomialFeatures適用前後の比較

一般線形回帰とPolynomialFeatures適用後の評価結果を比較し、派生変数の追加による性能の変化を確認する。

<div style="display:inline-block; min-width:800px; font-family:Consolas, 'Yu Gothic', monospace; line-height:27px; margin:4px 0 14px 0; color:#222;"><div style="background:#e7e9eb; padding:0 14px;"><strong><span style="color:#ff5a1f;">R²：0.976 → 0.995</span></strong></div><div style="background:#f4f5f6; padding:0 14px;"><strong>モデルの説明力が向上。</strong> R²は、モデルがデータの変動をどの程度説明できているかを表す指標。</div><div style="background:#e7e9eb; padding:0 14px;"><strong>1に近いほど、データの変動をよりよく説明できる。</strong></div><div style="background:#f4f5f6; padding:0 14px;">&nbsp;</div><div style="background:#e7e9eb; padding:0 14px;"><strong><span style="color:#ff5a1f;">MAE：3.00 → 1.28</span></strong></div><div style="background:#f4f5f6; padding:0 14px;">実測値と予測値の平均絶対誤差が、約3.00点から1.28点に減少。</div><div style="background:#e7e9eb; padding:0 14px;"><strong>値が小さいほど、平均的な予測誤差が小さい。</strong></div><div style="background:#f4f5f6; padding:0 14px;">Mean Absolute Error：平均絶対誤差</div><div style="background:#e7e9eb; padding:0 14px;">&nbsp;</div><div style="background:#f4f5f6; padding:0 14px;"><strong><span style="color:#ff5a1f;">MSE：9.55 → 2.19</span></strong></div><div style="background:#e7e9eb; padding:0 14px;">実測値と予測値の誤差を二乗して平均した値。</div><div style="background:#f4f5f6; padding:0 14px;"><strong>大きな誤差をより強く反映し、値が小さいほど良い。</strong></div><div style="background:#e7e9eb; padding:0 14px;">単位は元の目的変数の二乗単位。</div><div style="background:#f4f5f6; padding:0 14px;">Mean Squared Error：平均二乗誤差</div><div style="background:#e7e9eb; padding:0 14px;">&nbsp;</div><div style="background:#f4f5f6; padding:0 14px;"><strong><span style="color:#ff5a1f;">RMSE：3.09 → 1.48</span></strong></div><div style="background:#e7e9eb; padding:0 14px;">MSEの平方根を取った値。</div><div style="background:#f4f5f6; padding:0 14px;"><strong>元のデータと同じ単位で誤差を確認でき、値が小さいほど良い。</strong></div><div style="background:#e7e9eb; padding:0 14px;">Root Mean Squared Error：二乗平均平方根誤差</div></div>